pkg/scheduler/queue/queue_test.go

```diff
@@ -63,10 +78,149 @@ func makeSchedulerRequest(tenantID string) *SchedulerRequest {
+			promauto.With(nil).NewHistogram(prometheus.HistogramOpts{}),
+		)
+
+		ctx := context.Background()
+		require.NoError(t, queue.starting(ctx))
+		t.Cleanup(func() {
+			require.NoError(t, queue.stop(nil))
+		})
+
+		// fill queue first with the slow queries, then the normal queries
+		for _, queueDimensionFunc := range []func() []string{slowQueueDimensionFunc, normalQueueDimensionFunc} {
+			startProducersChan := make(chan struct{})
+			producersErrGroup, _ := errgroup.WithContext(ctx)
+
+			runProducer := runQueueProducerIters(
+				queue, maxQueriersPerTenant, totalRequests/2, numProducers, numTenants, startProducersChan, queueDimensionFunc,
+			)
+			for producerIdx := 0; producerIdx < numProducers; producerIdx++ {
+				producerIdx := producerIdx
+				producersErrGroup.Go(func() error {
+					return runProducer(producerIdx)
+				})
+			}
+			close(startProducersChan)
+			err := producersErrGroup.Wait()
+			if err != nil {
+				require.NoError(t, err)
+			}
+		}
+
+		// emulate delay when consuming the slow queries
+		consumeFunc := func(request Request) error {
+			schedulerRequest := request.(*SchedulerRequest)
+			if schedulerRequest.AdditionalQueueDimensions[0] == slowConsumerQueueDimension {
+				time.Sleep(slowConsumerLatency)
+			}
+
+			queueTime := time.Since(schedulerRequest.EnqueueTime)
+			additionalQueueDimensionLabels := strings.Join(schedulerRequest.AdditionalQueueDimensions, ":")
+			queueDuration.With(prometheus.Labels{"additional_queue_dimensions": additionalQueueDimensionLabels}).Add(queueTime.Seconds())
+			return nil
+		}
+
+		// consume queries
+		queueConsumerErrGroup, ctx := errgroup.WithContext(ctx)
+		startConsumersChan := make(chan struct{})
+		runConsumer := runQueueConsumerIters(ctx, queue, totalRequests, numConsumers, startConsumersChan, consumeFunc)
+
+		for consumerIdx := 0; consumerIdx < numConsumers; consumerIdx++ {
+			consumerIdx := consumerIdx
+			queueConsumerErrGroup.Go(func() error {
+				return runConsumer(consumerIdx)
+			})
+		}
+
+		close(startConsumersChan)
+		err := queueConsumerErrGroup.Wait()
+		if err != nil {
+			require.NoError(t, err)
+		}
+
+		// record total queue duration by queue dimensions and whether the queue splitting was enabled
+		for _, queueDimension := range []string{normalQueueDimension, slowConsumerQueueDimension} {
+			queueDurationTotals[additionalQueueDimensionsEnabled][queueDimension] = promtest.ToFloat64(
+				queueDuration.With(prometheus.Labels{"additional_queue_dimensions": queueDimension}),
+			)
+		}
+
+		promRegistry.Unregister(queueDuration)
+	}
+
+	// total or average time in queue for a normal queue item should be roughly cut in half
+	// when queue splitting is enabled, as the average normal queue item waits behind
+	// half of the slow queue items, instead of waiting behind all the slow queue items.
+	expected := queueDurationTotals[false][normalQueueDimension] / 2
+	actual := queueDurationTotals[true][normalQueueDimension]
```

Comment:
Given all we need here is the duration of normal requests and not the slow requests, could we simplify `queueDurationTotals` to only store the normal request durations?

Reply:
we could, but even though were are not asserting on it at the end, it is still a metric that is helpful to look at when building and understanding the behavior of the test case